src/main/java/com/redhat/devtools/lsp4ij/features/codeBlockProvider/LSPCodeBlockProvider.java

```diff
@@ -0,0 +1,201 @@
+            }
+
+            // Find containing text ranges that are bounded by brace pairs
+            CharSequence documentChars = document.getCharsSequence();
+            List<TextRange> containingTextRanges = new ArrayList<>(textRanges.size());
+            for (TextRange textRange : textRanges) {
+                if (textRange.getLength() > 1) {
+                    int startOffset = textRange.getStartOffset();
+                    int endOffset = textRange.getEndOffset();
+
+                    char startChar = documentChars.charAt(startOffset);
+                    char endChar = documentChars.charAt(endOffset - 1);
+
+                    // If aligned on an open brace and this ends with the expected close brace, use it
+                    if ((startOffset == openBraceOffset)) {
+                        if ((closeBraceChar != null) && (closeBraceChar == endChar)) {
+                            return textRange;
+                        } else {
+                            return null;
+                        }
+                    }
+                    // If aligned on a close brace and this starts with the expected open brace, use it
+                    else if (((endOffset - 1) == closeBraceOffset)) {
+                        if ((openBraceChar != null) && (openBraceChar == startChar)) {
+                            return textRange;
+                        } else {
+                            return null;
+                        }
+                    }
+                    // Otherwise see if it starts and ends with a known brace pair and we'll find the "closest" below
+                    else if ((openBraceOffset == -1) && (closeBraceOffset == -1) && LSPCodeBlockUtils.isCodeBlockStartChar(file, startChar)) {
+                        Character pairedCloseBraceChar = LSPCodeBlockUtils.getCodeBlockEndChar(file, startChar);
+                        if ((pairedCloseBraceChar != null) && (pairedCloseBraceChar == endChar)) {
+                            containingTextRanges.add(textRange);
+                        }
+                    }
+                }
+            }
+
+            // Return the closest (i.e., smallest) containing text range
+            if (!ContainerUtil.isEmpty(containingTextRanges)) {
+                containingTextRanges.sort(Comparator.comparingInt(TextRange::getLength));
+                return ContainerUtil.getFirstItem(containingTextRanges);
+            }
+        }
+
+        return null;
+    }
+
+    @Nullable
+    private static TextRange getUsingFoldingRanges(@NotNull PsiFile file,
+                                                   @NotNull Document document,
+                                                   int offset,
+                                                   @Nullable Character openBraceChar,
+                                                   @Nullable Character closeBraceChar) {
+        List<FoldingRange> foldingRanges = LSPFoldingRangeBuilder.getFoldingRanges(file);
```

Comment:
This is effectively the same logic that was used previously to derive code blocks from folding ranges.